spi.c

```diff
@@ -0,0 +1,103 @@
+/* 
+  Not part of GRBL, KeyMe specific
+*/
+
+#include "spi.h"
+
+#define SPI_TRANSFER_FLAG (SPSR & (1<<SPIF))
+
+
+//Setup spi
+void spi_init()
+{
+
+  DDRB = ((1<<DDB2)|(1<<DDB1) | (1 << DDB0)); //spi pins on port b MOSI SCK,SS outputs
+
+  SPCR = ((1<<SPE)|(1<<MSTR)|(1<<SPR0));//|(1<<CPOL)|(1<<CPHA));  // SPI enable, Master, f/16
+
+  //Configure SCS Pins as outputs
+  bit_true(SCS_DDR_PORT, 1 << SCS_XTABLE_DDR_PIN); 
+  bit_true(SCS_DDR_PORT, 1 << SCS_YTABLE_DDR_PIN); 
+  bit_true(SCS_DDR_PORT, 1 << SCS_GRIPPER_DDR_PIN); 
+  bit_true(SCS_DDR_PORT, 1 << SCS_CAROUSEL_DDR_PIN); 
+
+  //Set SCS to low for all steppers
+  bit_false(SCS_PORT, 1 << SCS_XTABLE_PIN);
+  bit_false(SCS_PORT, 1 << SCS_YTABLE_PIN);
+  bit_false(SCS_PORT, 1 << SCS_GRIPPER_PIN);
+  bit_false(SCS_PORT, 1 << SCS_CAROUSEL_PIN);
+
+}
+
+void spi_transact_array(uint8_t * dataout, uint8_t * datain, volatile uint8_t *cs_port, uint8_t cs_pin)
+{
+  bit_true(*cs_port, 1 << cs_pin);
+  uint8_t len = sizeof(dataout) / sizeof(uint8_t);
+
+  uint8_t idx;
+  for(idx = 0; idx < len; idx++) {
+    SPDR = dataout[idx];        //SPDR - SPI Data register
+    while(!SPI_TRANSFER_FLAG);  //Wait for transmit to be completed
+    datain[idx] = SPDR;
+  }
+  bit_false(*cs_port, 1 << cs_pin);
+
+}
+
+void spi_transmit_array(uint8_t * dataout, volatile uint8_t *cs_port, uint8_t cs_pin)
+{
+  bit_true(*cs_port, 1<<cs_pin);
+  uint8_t len = sizeof(dataout)/sizeof(uint8_t);
+
+  uint8_t idx;
+  for(idx = 0; idx < len; idx++) {
+    SPDR = dataout[idx];//SPDR - SPI Data register
+    uint16_t timeout = 0xFFFF;
+    while (!(SPI_TRANSFER_FLAG)) {
+      if (timeout == 0) {
+        bit_false(*cs_port, 1 << cs_pin);
+        return;
+      }
+    }
+  }
+  return;
+}
+
+
+uint8_t spi_transact_byte(uint8_t data, volatile uint8_t *cs_port, uint8_t cs_pin)
+{
+  bit_true(*cs_port, 1 << cs_pin);
+  *cs_port &= ~(1 << cs_pin);
+  SPDR = data;
+
+  uint16_t timeout = 0xFFFF;
+  while (!(SPI_TRANSFER_FLAG))  {
+    if (timeout == 0) {
+      bit_false(*cs_port, 1 << cs_pin);
+      return 0;
+    }
+  }
+  bit_false(*cs_port, 1 << cs_pin);
+  uint8_t ret = SPDR;
+
+  return ret;
+}
+
+void spi_transmit_byte(char data, volatile uint8_t *cs_port, uint8_t cs_pin)
+{
+  bit_true(*cs_port, 1 << cs_pin);
+  SPDR = data;
+  
+  uint16_t timeout = 0xFFFF;
+  while (!(SPI_TRANSFER_FLAG)) {
+    if (timeout == 0) {
+      bit_false(*cs_port, 1 << cs_pin);
+      return;
+    }
+  }
+
+  bit_false(*cs_port, 1 << cs_pin);
+  return;
+
```

Comment:
Transmit/transact functions are redundant.  A singular transact function based on transact_array should really be the bottom call, with wrapper functions performing parameter passthrough / translation.